jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/sso/SsoCredentialProvider.kt

```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials.sso
+
+import software.amazon.awssdk.auth.credentials.AnonymousCredentialsProvider
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sso.SsoClient
+import software.amazon.awssdk.utils.cache.CachedSupplier
+import software.amazon.awssdk.utils.cache.RefreshResult
+import java.time.Duration
+import java.time.Instant
+
+class SsoCredentialProvider(
+    ssoUrl: String,
+    ssoRegion: String,
+    private val ssoAccount: String,
+    private val ssoRole: String,
+    onPendingToken: NotifyTokenPending
+) : AwsCredentialsProvider {
+    private val client = SsoClient.builder()
+        .credentialsProvider(AnonymousCredentialsProvider.create())
+        .region(Region.of(ssoRegion))
+        .build()
+
+    private val accessTokenProvider = SsoAccessTokenProvider(ssoUrl, ssoRegion, onPendingToken)
+
+    private val sessionCache: CachedSupplier<SsoCredentialsHolder> = CachedSupplier.builder<SsoCredentialsHolder>(this::refreshCredentials).build()
+
+    override fun resolveCredentials(): AwsCredentials = sessionCache.get().credentials
+
+    private fun refreshCredentials(): RefreshResult<SsoCredentialsHolder> {
+        val roleCredentials = client.getRoleCredentials {
+            it.accessToken(accessTokenProvider.accessToken().accessToken)
```

Comment:
do we need to do something if the access token is no longer valid?

Reply:
The `.accessToken()` call should always return a valid one since it will deal with the refreshing